node_modules/oae-tenants/lib/api.js

```diff
@@ -132,15 +142,32 @@ var init = module.exports.init = function(_serverConfig, callback) {
  * Get a list of all available tenants from cache. The global admin tenant will be excluded from the resulting tenant list
  *
  * @param  {Boolean}        [excludeDisabled]   Whether or not disabled tenants should be included. By default, all tenants will be returned
+ * @param  {Number}         [start]             The index to start paging at, defaults to 0
+ * @param  {Number}         [limit]             The maximum number of tenants to return, if not specified all tenants will be returned
+ * @param  {String}         [q]                 Query string used to filter results
  * @return {Object}                             An object keyed by tenant alias holding all the tenants
  */
-var getTenants = module.exports.getTenants = function(excludeDisabled) {
+var getTenants = module.exports.getTenants = function(excludeDisabled, start, limit, q) {
+    excludeDisabled = OaeUtil.castToBoolean(excludeDisabled);
+    start = OaeUtil.getNumberParam(start, 0);
+    // Note that we don't apply a default limit here as some methods need ALL tenants
+    limit = OaeUtil.getNumberParam(limit);
+
     var filteredTenants = {};
-    _.each(tenants, function(tenant, tenantAlias) {
-        // Exclude all disabled tenants when `exludeDisabled` has been provided
-        if (!tenant.isGlobalAdminServer && (!excludeDisabled || tenant.active)) {
-            filteredTenants[tenantAlias] = _copyTenant(tenant);
-        }
+    // Filter out the admin tenant and disabled tenants if requested
+    var preFiltered = excludeDisabled? _.where(sortedTenants, {active: true, isGlobalAdminServer: false}): _.where(sortedTenants, {isGlobalAdminServer: false});
+    if (_.isString(q)) {
+        preFiltered = _.filter(preFiltered, function(tenant) {
+            return tenant.alias.indexOf(q) !== -1 || tenant.displayName.indexOf(q) !== -1 || tenant.host.indexOf(q) !== -1;
```

Comment:
I don't feel strong about lunr.js vs elasticsearch, as long as it perform well with 20k institutions.